working/1847 - FinalizationRegistry/proposal.md

```diff
@@ -0,0 +1,379 @@
+  /// or the same, finalization token.
+  ///
+  /// The callback may be called at most once per registration, and not
+  /// for registrations which have been unregistered since they were registered.
+  void register(F value, FT finalizationToken, {Object? unregisterToken});
+
+  /// Unregisters any finalization callbacks registered with [unregisterToken]
+  /// as unregister-token.
+  ///
+  /// After unregistering, those callbacks will not happen even if the
+  /// registered object becomes inaccessible.
+  void unregister(Object? unregisterToken);
+}
+
+/// Represents a weak reference to the given [target].
+///
+/// Weak references don't prevent garbage collection from reclaiming an object
+/// which is not reachable through any other (non-[WeakRef]) objects. If
+/// runtime system reclaims such _weakly reachable_ object then all [WeakRef]
+/// objects which reference it will be cleared to contain [null] as their
+/// target.
+abstract class WeakRef {
+  /// Create a [WeakRef] pointing to the given [target].
+  external factory WeakRef(Object? target);
+
+  /// Get the current object pointed by the [WeakRef].
+  external Object? get target;
+
+  /// Set the current object pointed by the [WeakRef].
+  external void set target (Object? newTarget);
+}
+```
+
+The following classes are added to `dart:ffi` library:
+
+```dart
+/// Any variable which has a static type that is a subtype of a [Finalizable]
+/// is guaranteed to be alive for the full duration of a scope in which it is
+/// defined.
+///
+/// In other words if an object is referenced by such a variable it is
+/// guaranteed to *not* be considered unreachable for the duration of the scope.
+abstract class Finalizable {
+  external factory Finalizable._();
+}
+
+typedef NativeFinalizer = Void Function(Pointer<Void>);
+typedef NativeFinalizerPtr = Pointer<NativeFunction<NativeFinalizer>>
+
+/// [FinalizationRegistry] which will execute its finalizers as early as
+/// possible without waiting for control to return to the event loop.
+///
+/// Will also invoke finalization callbacks when the isolate which created
+/// this finalization registry is shutting down.
+abstract class NativeFinalizationRegistry<F extends Finalizable>
+    extends FinalizationRegistry<F, Pointer> {
+  /// Creates a finalization registry with the given finalization
+  /// callback.
+  ///
+  /// Note: [callback] is expected to be a native function which can be
+  /// executed outside of a Dart isolate. This means that passing an FFI
+  /// trampoline (a function pointer obtained via [Pointer.fromFunction]) is
+  /// not supported for arbitrary Dart functions. This constructor will throw
+  /// if an unsupported [callback] is passed to it.
+  ///
+  /// [callback] might be invoked on an arbitrary thread and not necessary
+  /// on the same thread that created [FinalizationRegistry].
+  external factory FinalizationRegistry(NativeFinalizerPtr callback);
```

Comment:
This cannot work on Windows without having the user compile C code as well.

We should consider adding a second factory tailored to `WinHeapFree` so that packages targeting system APIs on Windows (such as `package:win32`) can use finalizers without adding any C code that has to be compiled.

```dart
abstract class NativeFinalizationRegistry<F extends Finalizable>
    extends FinalizationRegistry<F, Pointer> {
  external factory NativeFinalizationRegistry(NativeFinalizerPtr callback);

  // ... same documentation as other factory ....
  //
  // Invokes `winHeapFree(processHeap, /*flags=*/ 0, finalizationToken)`.
  external factory NativeFinalizationRegistryWindows(WinHeapFreePtr callback);
}
```

Alternatively we could think about an API that has the signature as a type argument, an argument index for to which argument the finalization token should be passed, and a list argument for all the other arguments.

```dart
abstract class NativeFinalizationRegistry<F extends Finalizable, C extends Function>
    extends FinalizationRegistry<F, Pointer> {
  external factory NativeFinalizationRegistry(Pointer<NativeFunction<C>> callback,
      int tokenArgumentIndex, List otherArguments);
}
```

Most FFI packages that target system APIs do not want to have any C code in their package, we should support that use case.

Reply:
> Most FFI packages that target system APIs do not want to have any C code in their package, we should support that use case.

I agree with that. I expanded the wording in the section to highlight that we consider that undesirable to have users write any C code.